kernel/reduction.ml

```diff
@@ -355,28 +366,45 @@ and eqappr cv_pb l2r infos (lft1,st1) (lft2,st2) cuniv =
     (* 2 constants, 2 local defined vars or 2 defined rels *)
     | (FFlex fl1, FFlex fl2) ->
       (try
-          let cuniv = conv_table_key infos.cnv_inf fl1 fl2 cuniv in
-           convert_stacks l2r infos lft1 lft2 v1 v2 cuniv
+         let cuniv = conv_table_key infos.cnv_inf fl1 fl2 cuniv in
+         convert_stacks l2r infos lft1 lft2 v1 v2 cuniv
        with NotConvertible | Univ.UniverseInconsistency _ ->
-           (* else the oracle tells which constant is to be expanded *)
+         (* else the oracle tells which constant is to be expanded *)
          let oracle = CClosure.oracle_of_infos infos.cnv_inf in
          let (app1,app2) =
+           let aux appr1 lft1 fl1 tab1 v1 appr2 lft2 fl2 tab2 v2 =
+             match unfold_reference infos.cnv_inf tab1 fl1 with
+             | Def def1 -> ((lft1, (def1, v1)), appr2)
+             | Primitive op when check_native_args op v1 ->
+               let c =
+                 match fl1 with ConstKey c -> c | _ -> assert false in
+               let rargs, a, nargs, v1 = get_native_args1 op c v1 in
+               ((lft1,
+                 whd_stack infos.cnv_inf tab1 a
+                   (Zupdate a::(Zprimitive(op,c,rargs,nargs)::v1))),
+                appr2)
```

Comment:
This unfold reference stuff is getting complicated and is repeated 6 times (twice here, then twice for Proj/Flex and twice for Flex/other), maybe it should be moved to a separate function.
Something like
~~~ocaml
let unfold_ref_with_args infos tab fl v =
  match unfold_reference infos tab fl with
  | Undef _ | OpaqueDef _ -> None
  | Def def -> Some (def, v)
  | Primitive op when check_native_args op v ->
    let c = match fl with ConstKey c -> c | _ -> assert false in
    let rargs, a, nargs, v = get_native_args1 op c v in
    Some (whd_stack infos tab a (Zupdate a::(Zprimitive(op,c,rargs,nargs)::v)))
~~~
then 
~~~ocaml
let aux appr1 lft1 fl1 tab1 v1 appr2 lft2 fl2 tab2 v2 =
  match unfold_ref_with_args infos.cnv_inf tab1 fl1 v1 with 
  | Some t1 -> ((lft1, t1), appr2)
  | None -> match unfold_ref_with_args infos.cnf_inf tab2 fl2 v2 with 
    | Some t2 -> (appr1, (lft2, t2))
    | None -> raise NotConvertible
~~~

Reply:
I think it just grew organically. We can always `[@inline]` if we think it's an issue.